ldms/src/sampler/zfs_zpool/zfs_zpool.c

```diff
@@ -0,0 +1,529 @@
+		rc = 6;
+	}
+
+	if (rc == 0) {
+		record_instance = ldms_record_alloc(sampler_base->set,
+						    index_store.vdev_rec_idx);
+
+		if (record_instance == NULL) {
+			log_fn(LDMSD_LDEBUG,
+			       SAMP
+			       ": ldms_record_alloc() failed, resizing metric set\n");
+			resize_metric_set();
+			record_instance = ldms_record_alloc(sampler_base->set,
+							    index_store.
+							    vdev_rec_idx);
+			if (record_instance == NULL)
+				rc = 2;
+		}
+	}
+
+	if (rc == 0) {
+		rc = ldms_list_append_record(sampler_base->set, list_handle,
+					     record_instance);
+
+		/* zpoolname    0 */
+		ldms_record_array_set_str(record_instance, zpool_metric_ids[0],
+					  pool_name);
+		/* zpool state  1 */
+		ldms_record_array_set_str(record_instance, zpool_metric_ids[1],
+					  zpool_state_to_name((vdev_state_t)
+							      vs->vs_state,
+							      (vdev_aux_t) vs->
+							      vs_aux));
+		/* total     2 */
+		ldms_record_set_u64(record_instance, zpool_metric_ids[2],
+				    vs->vs_space);
+		/* allocated 3 */
+		ldms_record_set_u64(record_instance, zpool_metric_ids[3],
+				    vs->vs_alloc);
+		/* free     4 */
+		ldms_record_set_u64(record_instance, zpool_metric_ids[4],
+				    vs->vs_space - vs->vs_alloc);
+		/* used     5 */
+		cap =
+		    (vs->vs_space ==
+		     0) ? 0 : (vs->vs_alloc * 10000 / vs->vs_space) / 100;
+		ldms_record_set_u64(record_instance, zpool_metric_ids[5], cap);
+
+		/* Here we call the get_pool_scan function to fill the rest of the
+		 * record */
+		rc = get_pool_scan_status(nvroot, pool_name, &record_instance);
+
+	}
+	return (0);
+}
+
+static int get_zpool_count(zpool_handle_t * zhp, void *data)
+{
+	int rc = 0;
+
+	if (zhp != NULL) {
+		zpool_list_len++;
+		zpool_close(zhp);
+	} else {
+		rc = 1;
+	}
+	return (rc);
+}
+
+static int get_pool_scan_status(nvlist_t * nvroot, const char *pool_name,
+				ldms_mval_t * record_instance)
+{
+	uint_t c;
+	int64_t elapsed;
+	uint64_t pass_exam, paused_time, rate;
+	pool_scan_stat_t *ps = NULL;
+	char *state[DSS_NUM_STATES] = {
+		"NONE",
+		"scanning",
+		"finished",
+		"canceled"
+	};
+/*	operation_types func;*/
+	const char *func;
+
+	(void)nvlist_lookup_uint64_array(nvroot,
+					 ZPOOL_CONFIG_SCAN_STATS,
+					 (uint64_t **) & ps, &c);
+
+	/*
+	 * ignore if there are no stats
+	 */
+	if (ps == NULL)
+		return (0);
+
+	/*
+	 * return error if state is bogus
+	 */
+	if (ps->pss_state >= DSS_NUM_STATES || ps->pss_func >= POOL_SCAN_FUNCS) {
+		if (complained_about_sync % 1000 == 0) {
+			fprintf(stderr, "error: cannot decode scan stats: "
+				"ZFS is out of sync with compiled zfs_zpool (ldms)");
+			complained_about_sync++;
+		}
+		return (1);
+	}
+
+	switch (ps->pss_func) {
+
+	case POOL_SCAN_NONE:
+		func = operation_types[FUNC_NOFUNCREQ];
+		break;
+	case POOL_SCAN_SCRUB:
+		func = operation_types[FUNC_SCRUB];
+		break;
+	case POOL_SCAN_RESILVER:
+		func = operation_types[FUNC_RESILVER];
+		break;
+#ifdef POOL_SCAN_REBUILD
```

Comment:
You can drop the `#ifdef` here.  Older versions of ZFS will just never return this value.